code-conventions.md

```diff
@@ -112,56 +154,79 @@ Your comments should look something like these:
 * Use assertions liberally: if you're sure something is not null, assert so to the compiler
   - Makes finding bugs easier for developers
+* Assertions must **not** have side-effects - they may be skipped in real environments
 * Avoid checking non-null values for nullability
   - Unless working around buggy addons, it is rarely necessary
   - This is why ignoring nullness errors is particularly dangerous
 
-## Portability
-* **Absolutely no NMS-reliant code**
-  - No matter *how* it is done, it belongs to addons
-  - Updating for new Minecraft versions is already painful enough
-  ```java
-  import net.minecraft.server.v1_13_1.*; // BAD
-  ```
-* Target Minecraft versions are written in README
-  - Skript **must** run on these MC versions, in one way or another
-* Supported server implementations are Spigot and Paper
-  - Paper-specific functionality is acceptable
-  - Glowstone also works, don't intentionally break it
-  - CraftBukkit does not work, and please don't waste your time in fixing it
-* Avoid referencing Material enum, as it changed significantly in Spigot 1.13
-  - Instead, use Aliases.javaItemType
-  ```java
-  Material type = Material.DIRT; // Bad
-  ItemType type = Aliases.javaItemType("dirt"); // Good
-  ```
-  - Exceptions: <code>Material.AIR</code> and <code>Material.STONE</code>
-    - Air is better way to represent "nothing" than null
-    - Stone can be used to get dummy <code>ItemMeta</code>
-* Do not refer Biome enum directly, because it changed in 1.9 *and* 1.13
-  ```java
-  Biome.MEGA_SPRUCE_TAIGA_HILLS // Bad - 1.8 edition
-  Biome.MUTATED_REDWOOD_TAIGA_HILLS // Bad - 1.9-1.12 edition
-  Biome.GIANT_SPRUCE_TAIGA_HILLS // Bad - 1.13 edition
-  // Yes, these are exactly same biomes. No, I have no idea why this is the case
-* Skript must run with assertations enabled; use them in your development environment
-  - JVM flag <code>-ea</code> is used to enable them
+### Assertions
+
+Skript must run with assertations enabled; use them in your development environment. \
+The JVM flag <code>-ea</code> is used to enable them.
+
+
+## Minecraft Features
+
+### Avoid Version-Specific Code
+
+Contributions must **never** rely on the variable `net.minecraft.server` package.
+```java
+import net.minecraft.server.v1_13_1.*; // BAD
+```
+This makes updating between versions excessively difficult, and requires us to write different code for all supported versions.
+If a feature is not accessible via the Bukkit API it should not be included in Skript.
+> For features requiring 'NMS' please consider making a third-party addon.
+
+### Support the Target Versions
+The target Minecraft versions are written in our README.
+
+Skript **must** run on these MC versions, in one way or another.
+If your contribution breaks compatibility for any of these versions we cannot accept it.
+
+Please try to make sure contributions are future-safe, to the best of your ability.
+Avoid targeting features of specific versions (e.g. via reflection.)
```

Comment:
This can be interpreted as, for example, syntax for the 1.18 freezing mechanics not being allowed

Reply:
Probably can be written better, feel free to change this :)